Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -28,49 +28,369 @@
 
 #include "Chrome.h"
 #include "ChromeClient.h"
+#include "ComposedTreeIterator.h"
 #include "Document.h"
+#include "DocumentType.h"
 #include "Editor.h"
-#include "FindOptions.h"
+#include "HTMLAudioElement.h"
+#include "HTMLIFrameElement.h"
+#include "HTMLImageElement.h"
+#include "HTMLMeterElement.h"
+#include "HTMLObjectElement.h"
+#include "HTMLProgressElement.h"
+#include "HTMLScriptElement.h"
+#include "HTMLStyleElement.h"
+#include "HTMLVideoElement.h"
+#include "Position.h"
 #include "SimpleRange.h"
+#include "TextBoundaries.h"
 #include "TextIterator.h"
 
 namespace WebCore {
 namespace FragmentDirectiveRangeFinder {
 
-static SimpleRange collapseIfRootsDiffer(SimpleRange&& range)
+enum class BoundaryPointIsAtEnd { No, Yes };
+enum class WordBounded : bool { No, Yes };
+
+static bool isSearchInvisible(const Node& node)
 {
-    return &range.start.container->rootNode() == &range.end.container->rootNode()
-        ? WTFMove(range) : SimpleRange { range.start, range.start };
+    // TODO: The computed value of its display property is none.
+    
+    // TODO: If the node serializes as void.
+    
+    if (is<HTMLIFrameElement>(node)
+        || is<HTMLImageElement>(node)
+        || is<HTMLMeterElement>(node)
+        || is<HTMLObjectElement>(node)
+        || is<HTMLProgressElement>(node)
+        || is<HTMLStyleElement>(node)
+        || is<HTMLScriptElement>(node)
+        || is<HTMLVideoElement>(node)
+        || is<HTMLAudioElement>(node))
+        return true;
+    
+    // TODO: Is a select element whose multiple content attribute is absent.
+    
+    return false;
 }
 
-static std::optional<SimpleRange> rangeOfString(const String& target, Document& document)
+static bool isNonSearchableSubtree(const Node& node)
 {
-    if (target.isEmpty())
+    const Node* traversingNode = &node;
+    while (traversingNode) {
+        if (isSearchInvisible(*traversingNode))
+            return true;
+        traversingNode = traversingNode->parentInComposedTree();
+    }
+    return false;
+}
+
+static const Node& nearestBlockAncestor(const Node& node)
+{
+    const Node* currentNode = &node;
+    while (currentNode) {
+        if (!currentNode->isTextNode() && currentNode->renderer() && currentNode->renderer()->style().isDisplayBlockLevel())
+            return *currentNode;
+        currentNode = currentNode->parentInComposedTree();
+    }
+    return node.document();
+}
+
+static std::optional<BoundaryPoint> getBoundaryPointAtIndex(unsigned index, Vector<RefPtr<Node>>& textNodeList, BoundaryPointIsAtEnd isEnd)
```

Comment:
`const Vector<RefPtr<Node>>& textNodeList`. Usually we don't use "get" in function names.

Reply:
Maybe `boundaryPointAtIndexInNodes`?

Reply:
Add a link to https://wicg.github.io/scroll-to-text-fragment/#get-boundary-point-at-index?

Reply:
Can we also make this function take a Vector<Ref<Text>>&?